src/snowcli/cli/nativeapp/manager.py

```diff
@@ -35,3 +57,101 @@ def deploy_root(self) -> Path:
+
+        # Does a stage already exist within the app pkg, or we need to create one?
+        # Using "if not exists" should take care of either case.
+        log.info("Checking if stage exists, or creating a new one if none exists.")
+        create_stage_if_not_exists_stmt = f"""
+            create stage if not exists {app_pkg}.{schema_and_stage}
+                encryption = (TYPE = 'SNOWFLAKE_SSE');"""
+        self.execute_query_with_role(create_stage_if_not_exists_stmt, role)
+
+        # Perform a diff operation and display results to the user for informational purposes
+        log.info(
+            f"Performing a diff between the Snowflake stage and your local deploy_root {deploy_root} directory."
+        )
+        diff: DiffResult = stage_diff(
+            Path(deploy_root), f"{app_pkg}.{schema_and_stage}"
+        )
+        log.info("Listing results of diff:")
+        print_result(
+            ObjectResult(
+                {
+                    "only on local": ", ".join(diff.only_local),
+                    "only on stage": ", ".join(diff.only_on_stage),
+                    "modified/unknown": ", ".join(diff.different),
+                    "identical": ", ".join(diff.identical),
+                }
+            )
+        )
+
+        # Upload diff-ed files to app pkg stage
+        log.info(
+            f"Uploading diff-ed files from your local {deploy_root} directory to the Snowflake stage."
+        )
+        sync_local_diff_with_stage(
+            role=role, diff_result=diff, stage_path=f"{app_pkg}.{schema_and_stage}"
+        )
+
+    def app_run(self) -> SnowflakeCursor:
+
+        # It is possible that snowflake.yml does not contain some of the optional fields,
+        # and there is no snowflake.local.yml present in the search_path.
+        # In this case, we need to use a generated local yml which provides those default values.
+        generated_local_yml_overrides = generate_local_override_yml(
+            self.definition_manager.project_definition
+        )
+
+        app_pkg = get_required_field_from_definition(
+            "package", "name", self.definition, generated_local_yml_overrides
+        )
+
+        app_pkg_role = get_required_field_from_definition(
+            "package", "role", self.definition, generated_local_yml_overrides
+        )
+
+        show_app_cursor = self.execute_query_with_role(
+            app_pkg_role, f"show application packages like {app_pkg}"
+        )
+
+        if show_app_cursor.rowcount == 0:
+            # Create an app pkg, with distribution = internal to avoid triggering security scan
+            log.info(f"Creating new application package {app_pkg} in account.")
+            self.execute_query_with_role(
+                app_pkg_role,
+                f"""
+                create application package {app_pkg}
+                    comment = {SPECIAL_COMMENT}
+                    distribution = internal
+            """,
```

Comment:
I just remembered that `use role` change role in session, so if you execute 2 queries like

```
_execute_query("use role public; select ...")
_execute_query("select ...")
```

this second query will be executer with public role too.

Reply:
For the role changing, my experience says this is pretty nasty for debugging. One thing to consider would be to remove chaning role in single query in favour of chaning role in context. For this this we can use context managers, for example:

```py
with use_role("my_new_role"):
   _execute_query("select 1")
   _execute_query("select 2")

 _execute_query("select 2")  # executed using previous role
```

WDYT?

Reply:
Like this idea a lot, we did the same thing in the system test suite (can probably even copy over that exact context manager if we don't have anything like it in SnowCLI yet)

Reply:
Ack to both - edited with context manager `use_role`. I was initially under the impression that session is discarded after each `_execute_query` and hence setting the role in one statement won't make it available the next time we do `_execute_query`.